Generate lightProbe from CubeCamera

nipmarsh opened the pull request:
Add a method to generate a lightProbe from a cubeCamera.
It assumes that the cubeCamera renderTarget is RGBA

vobu committed:
fix(WebGLTextures): iOS UIWebView OffscreenCanvas.getContext("2d")

cordova iOS (as of 5.0) still uses UIWebView,
which provides OffscreenCanvas,
also OffscreenCanvas.getContext("webgl"),
but not OffscreenCanvas.getContext("2d")!

vobu committed:
refactor: linting

klausw committed:
WebXR: request common reference spaces as features

WebXR's requestReferenceSpace only works if the corresponding feature
was requested at session creation time. For simplicity, just ask for
the interesting ones as optional features, but be aware that the
requestReferenceSpace call will fail if it turns out to be unavailable.
('local' is always available for immersive sessions and doesn't need to
be requested separately.)

Mugen87 committed:
BoxGeometry: Introduce ES6 classes and ES6-ES5 conversion for builds.

Commit:
Generate lightProbe from CubeCamera

Commit:
Generate lightProbe from CubeCamera

WestLangley commented:
Is there a reason why you do not pass in a `WebGLRenderTarget.texture` instead of a `CubeCamera`?

See how `WebGLBackground` handles this:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLBackground.js#L105-L107

nipmarsh commented:
I could have only pass the renderTarget but the method is named fromCubeCamera so semanthically this is better to pass the CubeCamera

Commit:
Fix THREE.SphericalHarmonics3 to SphericalHarmonics3 in jsm

Commit:
Fix THREE.SphericalHarmonics3 to SphericalHarmonics3 in jsm

Commit:
Add example

Commit:
Add example

Mugen87 commented:
Live demo: http://raw.githack.com/nipmarsh/three.js/lightprobegenerator_fromCubeCamera/examples/webgl_lightprobe_cubecamera.html

Commit:
add example to files.js

Commit:
add example to files.js

WestLangley commented:
>but the method is named fromCubeCamera

You can change the name, no? Can you reuse the existing method? See how `WebGLBackground` handles this.

WestLangley commented on examples/js/lights/LightProbeGenerator.js:
```diff
@@ -116,6 +116,105 @@ THREE.LightProbeGenerator = {
+				color.setRGB( data[ i ] / 255, data[ i + 1 ] / 255, data[ i + 2 ] / 255 );
+				// pixel coordinate on unit cube
+
+				if (renderer.gammaOutput) {
```
All light probes must be in linear space. It has nothing to do with render settings.

nipmarsh replied:
Yes and this is why it depends on the renderer gammaOutput, keep in mind that the LightProbes is computed from a cubeCamera. During the rendering stage of the cubemap, gamma correction is applied if renderer.gammaOutput is true.

nipmarsh replied:
The name is meaningful I think. I do not understand what is the matter with it
I have to read pixels back from the GPU to the CPU. The existing method is a little bit different and based on images. It seams a lot more easy for people who are not very confortable with CG to handle this seamlessly

nipmarsh replied:
> Your environment map is a Cornell box with colors of the form `[ 0, 1, 1 ]`, so converting to and from sRGB space does nothing. Hence it is difficult to test if your color-space is correct.

True, I will check this tomorrow with another one

WestLangley replied:
>During the rendering stage of the cubemap, gamma correction is applied if renderer.gammaOutput is true.

Right. And I have suggested repeatedly that you handle the cube camera at the application level and pass in `cubeCamera.renderTarget` to a properly-named method.

I do appreciate you doing this, however. :-)

In fact, it should be implemented on the GPU. Then we can auto-generate irradiance maps from any WebGLRenderTargetCube.

nipmarsh replied:
A fromCubeRenderTarget will be ok for you ?
Yes, it should be on the GPU too, there is another pull request for this https://github.com/mrdoob/three.js/pull/16152

WestLangley replied:
I feel pretty strongly that would be a better API. That is why I asked you if there was a reason why you didn't do it that way.

If you want, we can ask @mrdoob what he prefers. I don't want to waste your time.

nipmarsh replied:
This is indeed a matter of point of view, maybe @mrdoob can give it conceptual advice ;-)

WestLangley commented:
Your environment map is a Cornell box with colors of the form `[ 0, 1, 1 ]`, so converting to and from sRGB space does nothing. Hence it is difficult to test if your color-space is correct.

Commit:
use gammaToLinear for color correction and update example with sky cubemap

Commit:
use gammaToLinear for color correction and update example with sky cubemap

Mugen87 committed:
Merge remote-tracking branch 'upstream/dev' into dev34

Mugen87 committed:
package.json: Clean up

Mugen87 committed:
Examples: Inherit loaders from Loader I.

Commit:
Rename fromCubeCamera to fromCubeRenderTarget

Commit:
Rename fromCubeCamera to fromCubeRenderTarget

gogoend committed:
zh doc: Translate the inherit method and property section in some loader pages.

WestLangley commented:
You are getting there, but there is still something wrong with the way you handle colorspace.

These should produce the same light probe, and it should be in linear colorspace.
```js
lightProbe.copy( LightProbeGenerator.fromRenderTargetCube( renderer, cubeCamera.renderTarget ) );

lightProbe.copy( LightProbeGenerator.fromCubeTexture( cubeTexture ) );
```

WestLangley commented:
The texture files are not needed in this PR, right?

gogoend committed:
zh doc: Sync the zh doc list with en's.

WestLangley committed:
Added tonemapping

gogoend committed:
Zh doc: remove the old canvas example links.

WestLangley committed:
Properly group indirect diffuse and indirect specular terms

Commit:
Merge remote-tracking branch 'upstream/dev' into lightprobegenerator_fromCubeCamera

nipmarsh commented on examples/js/lights/LightProbeGenerator.js:
```diff
@@ -116,6 +116,105 @@ THREE.LightProbeGenerator = {
 
+	},
+
+	fromCubeRenderTarget: function ( renderer, renderTarget ) {
```
Can you explain what is wrong for you?

nipmarsh replied:
I think color correction in threejs is a little bit messy.
So i should simply assume that the renderTarget is linear. What do you think?
Color correction should be applied as a render pass if the user wants it.
It would be more simple to understand and all buffer would be in linear color space has it should be without wondering in which color space we are working on!

WestLangley replied:
I think you can follow the same pattern in your method I used in #17385 and decode to linear based on the encoding of the `WebGLRenderTargetCube`.

WestLangley replied:
I think the function signature should be
```js
fromRenderTargetCube: function ( renderer, renderTargetCube ) {
```

Commit:
Cleanup and white space formatting

Commit:
Cleanup and white space formatting

gogoend committed:
zh doc: Sync with the en doc.

mrdoob committed:
Merge pull request #17377 from WestLangley/dev_clearcoat_tonemapping

Clearcoat Example: Added tone mapping

yomboprime committed:
LightningStrike: Make plasma sphere use physical transparency

mrdoob committed:
Merge pull request #17378 from WestLangley/dev_pbr_indirect_specular

Physical Shader: properly group indirect diffuse and indirect specular terms

mrdoob committed:
Merge pull request #17380 from yomboprime/crystal_plasma

LightningStrike: Make plasma sphere use physical transparency

sam6321 committed:
Fix depthTexture types in WebGLRenderTarget.d.ts

Added depthTexture argument to WebGLRenderTargetOptions and changed depthTexture property on WebGLRenderTarget to be of DepthTexture type instead of Texture type.

gogoend committed:
zh docs: Sync the zh docs to the latest en doc.

gogoend committed:
zh doc: fix for the wrong tab in list.js

mrdoob committed:
Merge pull request #17387 from sam6321/patch-1

Fix depthTexture types in WebGLRenderTarget.d.ts

mrdoob committed:
Merge pull request #17389 from gogoend/zh_doc

zh doc: Sync with en doc and remove some old links.

mrdoob committed:
r109dev

mrdoob committed:
Updated builds.

Mugen87 committed:
Merge pull request #17363 from Mugen87/dev37

Examples: Inherit loaders from Loader I.

Mugen87 committed:
Examples: Inherit loaders from Loader II.

Mugen87 committed:
Merge pull request #17395 from Mugen87/dev37

Examples: Inherit loaders from Loader II.

Mugen87 committed:
Examples: Inherit from Loader III.

Mugen87 committed:
Merge pull request #17397 from Mugen87/dev37

Examples: Inherit from Loader III.

Mugen87 committed:
Examples: Inherit from Loader IV.

Mugen87 committed:
Merge pull request #17398 from Mugen87/dev37

Examples: Inherit from Loader IV.

chkvskyi committed:
add missing setFromObject method

aardgoose committed:
document length parameter behaviour

mrdoob committed:
Merge pull request #17401 from aardgoose/doc-ba

document length parameter behaviour of <type>BufferAttribute

mrdoob committed:
Merge pull request #17400 from demimgga/box-helper

Add missing setFromObject method

Mugen87 committed:
Docs: Fix link to example in TGALoader page.

Mugen87 committed:
Merge pull request #17402 from Mugen87/dev37

Docs: Fix link to example in TGALoader page.

Mugen87 committed:
CompressedTextureLoader: Refactor parsing.

Mugen87 committed:
Examples: Inherit from Loader V.

mrdoob committed:
Merge pull request #17403 from Mugen87/dev37

CompressedTextureLoader: Refactor parsing.

mrdoob committed:
Merge pull request #17405 from Mugen87/dev29

Examples: Inherit from Loader V.

Mugen87 committed:
DataTextureLoader: Refactor parsing.

Mugen87 committed:
Examples: Clean up.

Mugen87 committed:
Merge pull request #17407 from Mugen87/dev37

Examples: Clean up.

Mugen87 committed:
Examples: Fix TS files.

Mugen87 committed:
Merge pull request #17410 from Mugen87/dev37

Examples: Fix TS files.

Oletus committed:
Fix toggling receiveShadow without changing material

Fix the case where there are two objects with the same material in the scene and one has receiveShadow=true and the other has receiveShadow=false.

This also prevents possible extra shader compiles when the same material is reused with receiveShadow enabled or disabled in the same scene.

gkjohnson committed:
Rename SpecularMIPLevelNode.d..ts to SpecularMIPLevelNode.d.ts

Mugen87 committed:
Merge pull request #17411 from gkjohnson/patch-2

Rename SpecularMIPLevelNode.d..ts to SpecularMIPLevelNode.d.ts

Mugen87 committed:
Merge pull request #17406 from Mugen87/dev29

DataTextureLoader: Refactor parsing.

Mugen87 committed:
Update builds.

Mugen87 committed:
DragControls: Remove deprecated API and checks.

mrdoob committed:
Merge pull request #17276 from Mugen87/dev34

BoxGeometry: Introduce ES6 classes and ES6-ES5 conversion for builds.

mrdoob committed:
Updated package-lock.json.

mrdoob committed:
Rollup: Tweaked Buble transforms.

mrdoob committed:
Updated builds.

mrdoob committed:
Merge pull request #17413 from higharc/fix-receiveShadow

Fix toggling receiveShadow without changing material

mrdoob committed:
Removed unneeded example.

mrdoob committed:
Updated builds.

mrdoob committed:
Merge pull request #17416 from Mugen87/dev37

DragControls: Remove deprecated API and checks.

mrdoob committed:
Merge pull request #16781 from vobu/bugfix/offscreencanvas-getcontext2d-uiwebview

WebGLTextures: Ensure OffscreenCanvas provides drawing context.

mrdoob committed:
Updated builds.

Carnaux committed:
Update OBJLoader2.html

In the code in the documentation  is "event.detail.loaderRootNode", but it returns a object3d as in the live example.

Oletus committed:
Take receiveShadow flag into account for shadowMask

This fixes objects with MeshLambertMaterial that used to erroneously receive shadows.

Mugen87 committed:
Merge pull request #17421 from Carnaux/patch-2

Update OBJLoader2.html

Commit:
Fix a spelling error

Mugen87 committed:
Merge pull request #17427 from r00ster91/patch-1

Docs: Fix a spelling error

mrdoob committed:
Merge pull request #17423 from higharc/fix-receiveShadow-shadowMask

Take receiveShadow flag into account for shadowMask

mrdoob committed:
Merge pull request #17429 from klausw/webxr-spacefeatures

WebXR: request common reference spaces as features

mrdoob committed:
Updated builds.

fabryx92 committed:
Update Loading-3D-models.html

To add another troubleshooting step that could be useful. The problem I had in order to make the model visible was that it was hidden in the dark. Once I added and positioned a light source, the model appeared. Maybe someone else could have the same problem and could solve it more quickly than me.

mrdoob committed:
Merge pull request #17432 from fabryx92/patch-1

Update Loading-3D-models.html

mrdoob committed:
GLTFExporter: Clean up.

mrdoob committed:
GLTFExporter: Fixed alphaMode handling

Mugen87 committed:
Loader: Remove Handlers.

Mugen87 committed:
Tests: Clean up

Mugen87 committed:
TS: Clean up

Mugen87 committed:
DracoLoader: Clean up.

mrdoob committed:
Merge pull request #17433 from mrdoob/gltfexporter

GLTFExporter: Fixed alphaMode handling

Mugen87 committed:
Merge pull request #17436 from Mugen87/dev29

DracoLoader: Clean up.

Mugen87 committed:
LegacyGLTFLoader: Fix typo.

Mugen87 committed:
JSM: Update modules.

Mugen87 committed:
Merge pull request #17438 from Mugen87/dev29

JSM: Update modules.

mrdoob committed:
package.json: removed deprecated jsnext:main.

mrdoob committed:
package.json: changed repository format.

mrdoob committed:
package.json: added more keywords.

takahirox committed:
Remove unnecessary EventDispatcher methods definitions from Geometry.d.ts

takahirox committed:
Remove EventDispatcher inheritance from DirectGeometry.d.ts

Mugen87 committed:
Merge pull request #17440 from takahirox/CleanupGeometryTS

Cleanup Geometry and DirectGeometry TS

Mugen87 committed:
Examples: Remove usage of BufferAttribute.setArray().

mrdoob committed:
Merge pull request #17439 from Mugen87/dev29

Examples: Remove usage of BufferAttribute.setArray().

mrdoob committed:
Merge pull request #17434 from Mugen87/dev38

Loader: Remove Handlers.

Mugen87 committed:
PointerLockControls: Clean up.

Mugen87 committed:
Merge pull request #17444 from Mugen87/dev29

PointerLockControls: Clean up.

Mugen87 committed:
Update builds.

Commit:
Merge remote-tracking branch 'origin/lightprobegenerator_fromCubeCamera' into lightprobegenerator_fromCubeCamera

Commit:
Rename function prototype and use convertColorToLinear function

mrdoob commented:
I think this PR got messed up... 
https://github.com/mrdoob/three.js/pull/17282/files

nipmarsh commented:
> I think this PR got messed up...
> https://github.com/mrdoob/three.js/pull/17282/files

Ouch, i accidently merge instead of rebase dev branch, sorry...
I will redo a clean pull request and close this one

WestLangley commented:
@nipmarsh Would you be willing to redo a clean PR?

nipmarsh commented:
I already rebased the PR softly, is it still messed up?



> Le 13 oct. 2019 à 03:15, WestLangley <notifications@github.com> a écrit :
> 
> @nipmarsh <https://github.com/nipmarsh> Would you be willing to redo a clean PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/mrdoob/three.js/pull/17282?email_source=notifications&email_token=AEVJSXVF7QJJI4SYSD5BGSLQOJZCFA5CNFSM4IM6MMXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBCL2PA#issuecomment-541375804>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEVJSXQ7QNTQCJDM4SSKO53QOJZCFANCNFSM4IM6MMXA>.
>
